File: js-template/leetcode/editor/cn/rotate-image.js
/*
 * @lc app=leetcode.cn id=48 lang=javascript
 * @lcpr version=30404
 *
 * [48] 旋转图像
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    var n = matrix.length;
    for (let row = 0; row < n - 1; row++) {
        for (let col = row; col < n - 1 - row; col++) {
            let coords = [[row, col], [col, n - 1 - row], [n - 1 - row, n - 1 - col], [n - 1 - col, row]]
            let temp = matrix[coords[3][0]][coords[3][1]];
            matrix[coords[3][0]][coords[3][1]] = matrix[coords[2][0]][coords[2][1]];
            matrix[coords[2][0]][coords[2][1]] = matrix[coords[1][0]][coords[1][1]];
            matrix[coords[1][0]][coords[1][1]] = matrix[coords[0][0]][coords[0][1]];
            matrix[coords[0][0]][coords[0][1]] = temp;
        }
    }
};
// @lc code=end

// your test code here

rotate([[1,2,3],[4,5,6],[7,8,9]])
// rotate([[1,2],[3,4]])


/*
// @lcpr case=start
// [[1,2,3],[4,5,6],[7,8,9]]\n
// @lcpr case=end

// @lcpr case=start
// [[5,1,9,11],[2,4,8,10],[13,3,6,7],[15,14,12,16]]\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/rotate-image.cpp
/*
 * @lc app=leetcode.cn id=48 lang=cpp
 * @lcpr version=30404
 *
 * [48] 旋转图像
 */

#include <iostream>
#include <vector>
#include <string>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int limit = n / 2;
        for (int i = 0; i <= limit; ++i) {
            for (int j = i; j < n - 1 - i; ++j) {
                int temp = 0;
                int record = matrix[j][n-1-i];
                matrix[j][n-1-i]=matrix[i][j];
                temp = matrix[n-1-i][n-1-j];
                matrix[n-1-i][n-1-j] = record;
                record = temp;
                temp = matrix[n-1-j][i];
                matrix[n-1-j][i] = record;
                record = temp;
                matrix[i][j] = record;
            }
        }
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    vector<vector<int>> matrix = {{1,2,3},{4,5,6},{7,8,9}};
    solution.rotate(matrix);
}



/*
// @lcpr case=start
// [[1,2,3],[4,5,6],[7,8,9]]\n
// @lcpr case=end

// @lcpr case=start
// [[5,1,9,11],[2,4,8,10],[13,3,6,7],[15,14,12,16]]\n
// @lcpr case=end

 */

